contracts/other/notional/Join1155.sol

```diff
@@ -88,34 +120,76 @@ contract Join1155 is IJoin, ERC1155TokenReceiver, AccessControl() {
         return amount;        
     }
 
-    /// @dev Transfer `amount` `asset` to `user`
+    /// @dev Before maturity, transfer `amount` `asset` to `user`.
+    /// After maturity, withdraw if necessary, then transfer `amount.wmul(accrual)` `underlying` to `user`.
     function exit(address user, uint128 amount)
         external override
         auth
         returns (uint128)
     {
-        return _exit(user, amount);
+        if (block.timestamp < maturity) {
+            return _exit(user, amount);
+        } else {
+            if (accrual == 0) redeem(); // Redeem all fCash, switch to underlying join, set accrual.
+            return _exitUnderlying(user, uint256(amount).wmul(accrual).u128());
+        }
     }
 
     /// @dev Transfer `amount` `asset` to `user`
     function _exit(address user, uint128 amount)
         internal
+        beforeMaturity
         returns (uint128)
     {
-        ERC1155 token = ERC1155(asset);
         storedBalance -= amount;
-        token.safeTransferFrom(address(this), user, id, amount, "");
+        ERC1155(asset).safeTransferFrom(address(this), user, id, amount, "");
         return amount;
     }
 
+    /// @dev Transfer `amount` `underlying` to `user`
+    function _exitUnderlying(address user, uint128 amount)
+        internal
+        afterMaturity
+        returns (uint128)
+    {
+        storedBalance -= amount;
+        MinimalTransferHelper.safeTransfer(IERC20(underlying), user, amount);
+        return amount;
+    }
+
+    /// @dev Switch to an exit-only underlying Join, converting all fCash holdings to underlying in the process.
+    function redeem()
+        public
+        afterMaturity
+    {
+        // Build an action to withdraw all mature fCash into underlying, then withdraw.
+        IBatchAction.BalanceAction[] memory withdrawActions = new IBatchAction.BalanceAction[](1);
+        withdrawActions[0] = IBatchAction.BalanceAction({
+            actionType: IBatchAction.DepositActionType.None,
+            currencyId: currencyId,
+            depositActionAmount: 0,
+            withdrawAmountInternalPrecision: 0,
+            withdrawEntireCashBalance: true,
+            redeemToUnderlying: true
+        });
+
+        IBatchAction(asset).batchBalanceAction(address(this), withdrawActions);
+
+        uint256 underlyingBalance = IERC20(underlying).balanceOf(address(this));
+        accrual = underlyingBalance.wdiv(storedBalance); // There is a rounding loss here. Some wei will be forever locked in the join.
```

Comment:
Yikes! How will this affect users?

Reply:
As discussed, the rounding will not affect users. The underlying received cannot be evenly divided between holders with a set number of decimals, so we round down on `accrual` and up to one wei per user will remain in the contract.